internal/controller/collection/reconciler.go

```diff
@@ -0,0 +1,277 @@
+	for _, or := range oldRefs.Difference(refs).UnsortedList() {
+		o := &v1alpha2.Object{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: or.Name,
+			}}
+		if err := r.client.Delete(ctx, o); err != nil {
+			collection.Status.SetConditions(xpv1.ReconcileError(err))
+			if err := r.client.Status().Update(ctx, collection); err != nil {
+				return ctrl.Result{}, errors.Wrap(err, errStatusUpdate)
+			}
+			return ctrl.Result{}, errors.Wrap(err, fmt.Sprintf("cannot delete observed object %s", or.Name))
+		}
+	}
+	collection.Status.SetConditions(xpv1.ReconcileSuccess(), xpv1.Available())
+
+	// Update references only if more or less objects got matched since the last reconciliation.
+	if !refs.Equal(oldRefs) {
+		collection.Status.Objects = refs.UnsortedList()
+	}
+
+	return ctrl.Result{RequeueAfter: r.pollInterval()}, r.client.Status().Update(ctx, collection)
+}
+
+func observedObjectName(collection client.Object, matchedObject client.Object) (string, error) {
+	h := md5.New() //#nosec G401 -- used only for unique name generation
+	if _, err := fmt.Fprintf(h, "%s-%s", collection.GetName(), matchedObject.GetUID()); err != nil {
+		return "", err
+	}
+	id, err := uuid.FromBytes(h.Sum(nil))
+	return id.String(), err
+}
+
+func observedObjectPatch(name string, matchedObject unstructured.Unstructured, collection *v1alpha1.ObservedObjectCollection) (*unstructured.Unstructured, error) {
+	objectManifestTemplate := `{
+"kind": "%s",
+"apiVersion": "%s",
+"metadata": {
+  "name": "%s",
+  "namespace": "%s"
+}
+}`
+	manifest := fmt.Sprintf(objectManifestTemplate, matchedObject.GetKind(), matchedObject.GetAPIVersion(), matchedObject.GetName(), matchedObject.GetNamespace())
+	observedObject := &v1alpha2.Object{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: collection.APIVersion,
+					Kind:       collection.Kind,
+					Name:       collection.Name,
+					UID:        collection.UID,
+				},
+			},
+		},
+		Spec: v1alpha2.ObjectSpec{
+			ResourceSpec: xpv1.ResourceSpec{
+				ProviderConfigReference: &collection.Spec.ProviderConfigReference,
+				ManagementPolicies:      []xpv1.ManagementAction{xpv1.ManagementActionObserve},
+			},
+			ForProvider: v1alpha2.ObjectParameters{
+				Manifest: runtime.RawExtension{
+					Raw: []byte(manifest),
+				},
+			},
+		},
+	}
+	if t := collection.Spec.Template; t != nil {
+		if len(t.Metadata.Labels) > 0 {
+			observedObject.SetLabels(t.Metadata.Labels)
+		}
+		if len(t.Metadata.Annotations) > 0 {
+			observedObject.SetAnnotations(t.Metadata.Annotations)
+		}
+	}
+	v, err := runtime.DefaultUnstructuredConverter.ToUnstructured(observedObject)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot convert to unstructured")
+	}
+	u := &unstructured.Unstructured{Object: v}
+	u.SetGroupVersionKind(v1alpha2.ObjectGroupVersionKind)
+	u.SetName(observedObject.Name)
+	return u, nil
```

Comment:
Do we have to use unstructured client to work with `Object` type? I believe we could just work with the `Object` type without this conversion.

Reply:
We need to use unstructured because we create Object using server-side apply.

Reply:
Correct me if I am wrong but `r.client.Patch` with `client.Apply` should work with typed objects as well. We need unstructured only if we pass partial objects (missing some required fields) ?

Reply:
FWIW my understanding matches @pedjak's - today the reality is that if you want to use SSA you need to use `Unstructured` objects.